SDL/Quintuplet.cu

```diff
@@ -602,138 +635,394 @@ __device__ bool SDL::passT5RZConstraint(struct SDL::modules& modulesInGPU, struc
+    float z_init=mdsInGPU.anchorZ[thirdMDIndex]/100;
+    float rt_init=mdsInGPU.anchorRt[thirdMDIndex]/100; //use the second MD as initial point
+
+    if (moduleType3==1)  // 1: if MD3 is in 2s layer
     {
-        slope = (z2 -z1)/(rt2 - rt1);
+        x_init=mdsInGPU.anchorX[secondMDIndex]/100;
+        y_init=mdsInGPU.anchorY[secondMDIndex]/100;
+        z_init=mdsInGPU.anchorZ[secondMDIndex]/100;
+        rt_init=mdsInGPU.anchorRt[secondMDIndex]/100;
     }
-    else
+
+    //start from a circle of inner T3.
+    // to determine the charge
+    int charge=0;
+    float slope3c=(y3-y_center)/(x3-x_center);
+    float slope1c=(y1-y_center)/(x1-x_center);
+    if((y3-y_center)>0 && (y1-y_center)>0) 
+    {
+        if (slope3c>slope1c) charge=-1; 
+        else if (slope3c<slope1c) charge=1;
+        if (slope1c>0 && slope3c<0) charge=-1;
+        if (slope1c<0 && slope3c>0) charge=1;
+    }
+    else if((y3-y_center)<0 && (y1-y_center)<0) 
+    {
+        if (slope3c>slope1c) charge=-1; 
+        else if (slope3c<slope1c) charge=1;
+        if (slope1c<0 && slope3c>0) charge=1;
+        if (slope1c>0 && slope3c<0) charge=-1;
+    }
+    else if ((y3-y_center)<0 && (y1-y_center)>0)
+    {
+        if ((x3-x_center)>0 && (x1-x_center)>0) charge = 1;
+        else if ((x3-x_center)<0 && (x1-x_center)<0) charge = -1;
+    }
+    else if ((y3-y_center)>0 && (y1-y_center)<0)
     {
-        slope = (z3 - z1)/(rt3 - rt1);
+        if ((x3-x_center)>0 && (x1-x_center)>0) charge = -1;
+        else if ((x3-x_center)<0 && (x1-x_center)<0) charge = 1;
     }
-    float residual4 = (layer4 <= 6)? ((z4 - z1) - slope * (rt4 - rt1)) : ((rt4 - rt1) - (z4 - z1)/slope);
-    float residual5 = (layer4 <= 6) ? ((z5 - z1) - slope * (rt5 - rt1)) : ((rt5 - rt1) - (z5 - z1)/slope);
 
-    // creating a chi squared type quantity
-    // 0-> PS, 1->2S
-    residual4 = (moduleLayer4 == 0) ? residual4/2.4f : residual4/5.0f;
-    residual5 = (moduleLayer5 == 0) ? residual5/2.4f : residual5/5.0f;
+    float pseudo_phi = atan((y_init-y_center)/(x_init-x_center)); //actually represent pi/2-phi, wrt helix axis z
+    float Pt=inner_pt, Px=Pt*abs(sin(pseudo_phi)), Py=Pt*abs(cos(pseudo_phi));
 
-    const float RMSE = sqrtf(0.5 * (residual4 * residual4 + residual5 * residual5));
+    if (x_init>x_center && y_init>y_center) //1st quad
+    {
+        if (charge==1) Py=-Py;
+        if (charge==-1) Px=-Px;
+    }
+    if (x_init<x_center && y_init>y_center) //2nd quad
+    {
+        if (charge==-1) {Px=-Px; Py=-Py;}
+    }
+    if (x_init<x_center && y_init<y_center) //3rd quad
+    {
+        if (charge==1) Px=-Px;
+        if (charge==-1) Py=-Py;
+    }        
+    if (x_init>x_center && y_init<y_center) //4th quad
+    {
+        if (charge==1) {Px=-Px; Py=-Py;}
+    }
+
+    if (moduleType3==0){
+        if (x4<x3 && x3<x2) Px=-abs(Px);
+        if (x4>x3 && x3>x2) Px=abs(Px);
+        if (y4<y3 && y3<y2) Py=-abs(Py);
+        if (y4>y3 && y3>y2) Py=abs(Py);
```

Comment:
These are also the special cases. Line 695 gives the general expression of momentum, Px and Py. Line 697 - Line 720 discusses about some special cases the Px and Py need to flip sign according to the center of the circle position and the charge. Line 724 to line 727 discuss about more rare cases, which are not included in above and we need to set the the momentum sign hard by hand.

Reply:
It looks like there a bit too many "special cases"? Are we sure there is no logic behind those, so that we can simplify?

If we can't avoid them, would it at least make sense to gather all of the "special cases" together in the code? And, in any case, could we have a (few) exaplanatory comment(s)?

Reply:
I think these special cases comes from the momentum sign assignment. From 20 lines above, I defined the momentum  sign by separating the circle into 4 quarters and compare geometrically the circle on the x-y plane is clockwise or anti-clockwise. And these cases are T5s that cross the edge of the 4 quarters(i.e. cross axis to separate the quarters). I'm afraid I can't think of a more clever way to simplify it... but I have added more comments in the code and if that can help